text/1444-union.md

```diff
@@ -178,39 +190,489 @@ fn is_zero(v: Value) -> bool {
+For `#[repr(C)]` unions this property is guaranteed by ISO C standards.
+
+### Dynamically sized fields
+
+Unions cannot contain fields with dynamically sized types like `[u8]` or
+`Trait`.
 
-#[test]
+### Unions and `#[derive]`
+
+Unions support deriving a few traits that do not require accessing their fields.
+Supported traits are
+
+- `Copy`, supported only for unions with `Copy` fields
+- `Clone`, supported only for unions implementing `Copy`, which can be cloned
+trivially
+- `Eq` (but not `PartialEq`), supported for unions with `Eq` fields.
+
+Since accessing union fields reliably requires extra knowledge, traits trying to
+do it (e.g. `PartialEq`) cannot be derived automatically.
+
+## Constructing union values
+
+A value of a union type can be created using the same syntax that is used for
+struct types, except that it must specify exactly one field:
+
+```rust
+let u = MyUnion { f1: 1 };
+```
+
+This expression will create a value of type `MyUnion` with active field `f1`.  
+Union expressions, unlike struct expressions, do not support functional record
+update (FRU) `MyUnion { f1: 1, ..v }`.  
+Note that creating a union with union expression is safe (no potentially
+unreliable data is accessed) and therefore doesn't require `unsafe` block.
+
+Alternatively, a union can be created in uninitialized state and filled in
+later.
+```rust
+let u: MyUnion;
+u.f1 = 1; // u becomes initialized, f1 is the active field
+          // NB: this is not actually implemented right now, but it should be
+```
+
+## Accessing union fields
+
+### Field access (`.`)
+
+Union fields can be accessed using the same dot syntax as struct fields
+```rust
+unsafe {
+    let a = u.f1;
+    u.f2 = 1.0;
+    fn_call(&u.f2);
+}
+```
+
+All accesses to union fields (both reads and writes, and also borrows) require
+`unsafe` blocks.
+The active field/fragment of a union can always be accessed safely, but
+reads of inactive fields/fragments can result in undefined behavior (see
+[Guarantees and undefined behavior](#guarantees-and-undefined-behavior) for
+more detail).
+
+Write to a union field can potentially overwrite contents of its other fields.  
+Write to a union field should not modify any bits outside of the modified field
+(except for possibly by running destructor of the old field value).
+
+When a new value is assigned to a union field, the old value of the field is
+dropped normally. This matches the behavior of `struct` fields, but differs
+from assignments to the whole union, which ignore field destructors by default
+(see [Unions and drop](#unions-and-drop) for more detail). Example:
+
+```rust
+struct NoisyDrop;
+impl Drop for NoisyDrop {
+    fn drop(&mut self) {
+        println!("BOOM!");
+    }
+}
+
+union U {
+    a: NoisyDrop
+}
+
+fn main() {
+    unsafe {
+        let mut u = U { a: NoisyDrop };
+        u.a = NoisyDrop; // BOOM!
+        u = U { a: NoisyDrop }; // Silence
+    }
+}
+```
```

Comment:
Calls `NoisyDrop2::drop` if `u` contains a valid value of `NoisyDrop2`, UB otherwise.
I will add this example to the text.